File: libs/php/SchemaLoader.php
<?php

require_once __DIR__ . '/Errors.php';

class IWXSchemaLoader
{
    private static array $registry = [];
    private const ALLOWED_TYPES = ['string', 'number', 'integer', 'boolean', 'date', 'object'];

    public static function registerSchema(array $schema): array
    {
        self::validateSchema($schema);
        self::$registry[$schema['schema_id']] = $schema;
        return $schema;
    }

    public static function getSchema(int $schemaId): array
    {
        if (!isset(self::$registry[$schemaId])) {
            throw new IWXError(IWXErrorCodes::SCHEMA_NOT_FOUND, 'Unknown IWX schema: ' . $schemaId);
        }
        return self::$registry[$schemaId];
    }

    public static function loadSchemaFile(string $path, ?callable $yamlParser = null): array
    {
        $ext = strtolower(pathinfo($path, PATHINFO_EXTENSION));
        $contents = file_get_contents($path);
        if ($contents === false) {
            throw new IWXError(IWXErrorCodes::SCHEMA_INVALID, 'Unable to read schema file', ['path' => $path]);
        }
        if ($ext === 'json') {
            $schema = json_decode($contents, true);
        } elseif ($ext === 'yml' || $ext === 'yaml') {
            if ($yamlParser === null) {
                throw new IWXError(
                    IWXErrorCodes::SCHEMA_INVALID,
                    'YAML schema provided but no yamlParser supplied. Pass a callable (e.g. yaml_parse) or pre-convert to JSON.',
                    ['path' => $path]
                );
            }
            $schema = $yamlParser($contents);
        } else {
            throw new IWXError(IWXErrorCodes::SCHEMA_INVALID, 'Unsupported schema file extension', ['path' => $path]);
        }
        return self::registerSchema($schema);
    }

    public static function loadSchemasFromDirectory(string $directory, ?callable $yamlParser = null): array
    {
        $loaded = [];
        foreach (scandir($directory) as $entry) {
            if ($entry === '.' || $entry === '..') {
                continue;
            }
            $path = rtrim($directory, DIRECTORY_SEPARATOR) . DIRECTORY_SEPARATOR . $entry;
            if (!is_file($path)) {
                continue;
            }
            $ext = strtolower(pathinfo($path, PATHINFO_EXTENSION));
            if (!in_array($ext, ['json', 'yml', 'yaml'], true)) {
                continue;
            }
            $loaded[] = self::loadSchemaFile($path, $yamlParser);
        }
        return $loaded;
    }

    private static function validateSchema(array $schema): void
    {
        if (!isset($schema['schema_id'], $schema['fields']) || !is_array($schema['fields'])) {
            throw new IWXError(IWXErrorCodes::SCHEMA_INVALID, 'Schema must include schema_id and fields[]');
        }
        $positions = [];
        foreach ($schema['fields'] as $field) {
            if (!isset($field['pos']) || !is_int($field['pos'])) {
                throw new IWXError(IWXErrorCodes::SCHEMA_INVALID, 'Field is missing numeric pos', ['field' => $field]);
            }
            if (in_array($field['pos'], $positions, true)) {
                throw new IWXError(IWXErrorCodes::SCHEMA_INVALID, 'Duplicate field position', ['pos' => $field['pos']]);
            }
            $positions[] = $field['pos'];
            if (empty($field['frontend']) || empty($field['backend'])) {
                throw new IWXError(IWXErrorCodes::SCHEMA_INVALID, 'Field requires frontend and backend names', ['field' => $field]);
            }
            if (isset($field['type']) && !in_array($field['type'], self::ALLOWED_TYPES, true)) {
                throw new IWXError(IWXErrorCodes::SCHEMA_INVALID, 'Unsupported field type', ['field' => $field]);
            }
        }
    }
}

// preload built-in schema
IWXSchemaLoader::registerSchema([
    'schema_id' => 1,
    'name' => 'Sale',
    'version' => 1,
    'fields' => [
        ['pos' => 0, 'backend' => 'tenant_id', 'frontend' => 'tenant', 'type' => 'string', 'required' => true],
        ['pos' => 1, 'backend' => 'date', 'frontend' => 'sale_date', 'type' => 'date', 'required' => true],
        ['pos' => 2, 'backend' => 'total', 'frontend' => 'amount', 'type' => 'number', 'required' => true],
    ],
]);

File: libs/python/schema_loader.py
import json
from pathlib import Path
from typing import Callable, Iterable
from .errors import ERROR_CODES, IWXError

ALLOWED_TYPES = {'string', 'number', 'integer', 'boolean', 'date', 'object'}
_registry = {}

BASE_DIR = Path(__file__).resolve().parents[2]
DEFAULT_SCHEMA_PATH = BASE_DIR / 'schemas' / 'sale' / 'schema.json'

def _load_default_schemas():
    if DEFAULT_SCHEMA_PATH.exists():
        with DEFAULT_SCHEMA_PATH.open('r', encoding='utf-8') as fh:
            schema = json.load(fh)
            register_schema(schema)


def _validate_schema(schema: dict):
    if not isinstance(schema, dict) or 'schema_id' not in schema or 'fields' not in schema:
        raise IWXError(ERROR_CODES['SCHEMA_INVALID'], 'Schema must include schema_id and fields[]')
    seen_positions = set()
    for field in schema['fields']:
        pos = field.get('pos')
        if pos is None or not isinstance(pos, int):
            raise IWXError(ERROR_CODES['SCHEMA_INVALID'], 'Field is missing numeric pos', {'field': field})
        if pos in seen_positions:
            raise IWXError(ERROR_CODES['SCHEMA_INVALID'], 'Duplicate field position', {'pos': pos})
        seen_positions.add(pos)
        if not field.get('frontend') or not field.get('backend'):
            raise IWXError(ERROR_CODES['SCHEMA_INVALID'], 'Field requires frontend and backend names', {'field': field})
        ftype = field.get('type')
        if ftype and ftype not in ALLOWED_TYPES:
            raise IWXError(ERROR_CODES['SCHEMA_INVALID'], 'Unsupported field type', {'field': field})


def register_schema(schema: dict):
    _validate_schema(schema)
    _registry[schema['schema_id']] = schema
    return schema


def get_schema(schema_id: int) -> dict:
    try:
        return _registry[schema_id]
    except KeyError as exc:
        raise IWXError(ERROR_CODES['SCHEMA_NOT_FOUND'], f'Unknown IWX schema: {schema_id}') from exc


def list_schemas():
    return list(_registry.values())


def load_schema_file(path: Path, yaml_loader: Callable[[str], dict] | None = None) -> dict:
    ext = path.suffix.lower()
    text = path.read_text(encoding='utf-8')
    if ext == '.json':
        schema = json.loads(text)
    elif ext in {'.yml', '.yaml'}:
        if yaml_loader is None:
            raise IWXError(
                ERROR_CODES['SCHEMA_INVALID'],
                'YAML schema provided but no yaml_loader supplied. Pass a callable like yaml.safe_load or pre-convert to JSON.',
                {'path': str(path)},
            )
        schema = yaml_loader(text)
    else:
        raise IWXError(ERROR_CODES['SCHEMA_INVALID'], 'Unsupported schema file extension', {'path': str(path)})
    return register_schema(schema)


def load_schemas_from_dir(directory: Path | str, yaml_loader: Callable[[str], dict] | None = None) -> Iterable[dict]:
    dir_path = Path(directory)
    loaded = []
    for file_path in dir_path.iterdir():
        if not file_path.is_file():
            continue
        if file_path.suffix.lower() not in {'.json', '.yml', '.yaml'}:
            continue
        loaded.append(load_schema_file(file_path, yaml_loader=yaml_loader))
    return loaded


_load_default_schemas()
